### PR TITLE
[BEAM-5074][SQL] improve BeamSqlRowComparator

### PR DESCRIPTION
**Pull request by amaliujia**
Refactoring BeamSqlRowComparator a bit to make it can return comparison results in either forward order or reversed order.

In the past, BeamSqlRowComparator only returned comparison results in reversed order. It was because BeamSortRel only supports `ORDER BY LIMIT` by using `Top` (`Top` by default returns elements in descending order). For example, if there is a query `SELECT * FROM table ORDER BY id ASC LIMIT 1`,  and assuming there are two rows in table with id `1` and `2` separately, in order to return the right result by using `Top`,  `BeamSqlRowComparator` have to tell `Top` that the row with id `1` is bigger (which means comparator returns in reversed order).

In order to support [ORDER BY ASC and ORDER BY DESC](https://issues.apache.org/jira/browse/BEAM-5064), there is a need to have a comparator that returns in forward order (if 1 < 2, then returns -1 to indicate 1 < 2). So in this PR,  BeamSqlRowComparator is changed to achieve this goal. 

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | --- | --- | --- | ---

### CODE REVIEW

**Comment by amaliujia** [comment]
R: @XuMingmin @akedin @apilloud

**Comment by apilloud** [comment]
I'm confused. Isn't the existing `orientation` field the same as `isReversed`? How does `isReversed` get hooked up?

**Comment by amaliujia** [comment]
@apilloud 

`orientation` tells what should be the order on a field. If the query has `ASC` on a field, then `orientation` is true, and if the query has `DESC`, then `orientation` is false.

However, if you observe what the code is written for `BeamSqlRowComparator` ([code linik](https://github.com/apache/beam/blob/master/sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamSortRel.java#L320)), you can find this line:
```
fieldRet *= (orientation.get(i) ? -1 : 1);
```

The comparator will reverse the comparison result. For example, if it is `ASC` and `orientation` is true, and if fieldRet was `-1`, then fieldRet will become 1, which means smaller value will be put after the bigger value (the reversed order). 


The reason this implementation is correct, as I tried to explain in the PR description, is that `ORDER BY LIMIT` uses `Top.of`. `Top` transform returns largest elements first (another reversed order), which finally made the SQL queries' results correct (ASC gives increasing order result and DESC gives decreasing order result). 


The `isReversed`, in stead, allows we control the behavior of `BeamSqlRowComparator`.  The comparator will return -1 for `ASC` and 1.compareTo(2) if `isReversed = false`, and 1 otherwise. Right now, the comparator returns 1 for `ASC` and 1.compareTo(2).

It will be useful if I want to implement `ORDER BY ASC/DESC`

**Comment by apilloud** [comment]
We discussed this. You are going to post the code that requires this change and we can discuss this in the context of that.

**Comment by amaliujia** [comment]
@apilloud 

Per discussion, I updated this PR. PTAL.

**Comment by amaliujia** [comment]
run java precommit

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamSortRel.java [hunk 1]
```diff
@@ -317,12 +317,31 @@ public int compare(Row row1, Row row2) {
...
         if (fieldRet != 0) {
           return fieldRet;
         }
       }
       return 0;
     }
   }
+
+  private static class ReversedBeamSqlRowComparator implements Comparator<Row>, Serializable {
+    private BeamSqlRowComparator comparator;
```
Comment (apilloud) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamSortRel.java]: nit: you should make this `final`.

File: sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamSortRel.java [hunk 2]
```diff
@@ -317,12 +317,31 @@ public int compare(Row row1, Row row2) {
...
+
+    public ReversedBeamSqlRowComparator(
+        List<Integer> fieldsIndices, List<Boolean> orientation, List<Boolean> nullsFirst) {
+      comparator = new BeamSqlRowComparator(fieldsIndices, orientation, nullsFirst);
+    }
+
+    @Override
+    public int compare(Row row1, Row row2) {
+      int ret = comparator.compare(row1, row2);
```
Comment (apilloud) [inline sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/impl/rel/BeamSortRel.java]: nit: You can simplify this function as `return comparator.compare(row2, row1);`

**Comment by apilloud** [comment]
run java precommit

**Pull request merged**